### PR TITLE
Implement Builders

### PR DESCRIPTION
**Pull request by Emxx52**
1. Builder bots base implementation:
   - added WheeledBuilder, TrackedBuilder, WingedBuilder, LeggedBuilder;
2. neutron gun moves and stuff;
3. new icons

TODO:
* [x] research;
* [x] documentation;
* [x] visual improvements (more particles);
* [x] building while flying
* [x] interface
* [x] new Code Battles
* [x] dev-targetpractice #1058

### CODE REVIEW

**Comment by krzys-h** [comment]
There are some important things to be addressed before merging this:

1. Code battles are completly unplayable now. The main idea behind this game mode was to make everything happen with just one robot and one Titanium cube. How do we do that now without a robot that can both move things and build?

2. Up until now, it was possible to complete a lot of the missions using code (well, at least in theory). After these changes, it will be impossible because you don't have any robots that can build at the start of a mission.

**Comment by tomangelo2** [comment]
1. In order to make it finishable we either need to add second bot or place factory somewhere near. Both will slightly collide with main idea of this mode.
From the other hand Grabbers with build() function is a gameplay change (unless we won't deny a possibility to use it in Missions). But then we have something like this on our page: 

> Also, it is important to know that Colobot: Gold Edition is just a remake. Any sophisticated features like new objects, [...] will not be included.

Won't these bots be kind of opposition of this sentence?
I guess we kinda trapped ourselves at this moment.

**Comment by krzys-h** [comment]
Our implementation of code battles was NEVER based on CeeBot, it developed as a completly separate idea. We didn't even know about their existence there at the time. I know you can just start with multiple bots, but the idea was to make it start with just one program that does everything (and it greatly simplifies setting up matches and sharing programs)

Builder bots might have been a part of the story, but it was never said clearly - we just imply it from the fact that previous expeditions had buildings. And it doesn't change the fact that it has a major impact on gameplay, which we were meant to avoid.

I don't have anything against implementing BuilderBots for the purposes of new userlevels, but I'd rather keep the main game as it is now.

**Comment by tomangelo2** [comment]
I don't remember BuilderBots anywhere in the original story, there were some wrecks with neutron gun-like tool, but these would be as well previous revision of cannon bots (DefenseTower use similar thing to shoot Aliens). What built all these buildings on First Expedition - we don't know, they just were. We can make theories about BuilderBots that wasn't included for Second Expedition because NASA said so, or it was Grabbers with removed neutron gun in SE because NASA said so (we could also tell that robot from 4-3 wasn't different than SE equivalent, without visible neutron gun, some might still say it was hidden), or they was prebuilded, delivered in parts and assembled from scratch, but I think there is no bigger official story behind these ruins, originally they just were placed as a decorations without any backstory.

**Comment by MrSimbax** [comment]
Are you **SERIOUSLY** ok with **Grabbers** being able to build? It was supposed to be a **placeholder** until we get the builder bots, now we have builder bots, and suddenly the Grabbers was the plan all along? Are you kidding me?

The breaking of the code battles is not the issue here, the issue here is you don't want to IMPROVE the new gamemode so that it's like it should be from the beginning (starting with multiple bots => much more possibilities!), you want to keep it in this alpha stage for next releases and the final release just because you suddenly changed your idea of the new game mode.

Merge this pull request and improve the code battles or remove them from the game altogether if you don't care about the quality of this game and just want to pump new broken half-done features and stay in alpha stage forever.

The missions in-game will be reworked anyway, there is nothing to discuss here about that.

Notice that didn't even touch the plot and I fail to see how it's in any way related to the pull request.

**Comment by krzys-h** [comment]
Grabbers were supposed to be a placeholder, but since that placeholder was in the game for 4 years or so it already kind of became a feature.

You can make a code battle level that starts with multiple bots, that's not the problem. The problem is that after these changes it is not possible to make a code battle that starts with just one.

Removing the possibility to complete at least some of the missions with code until we finally decide to rework them (which will probably take at least another 4 years...) seems like a huge step back to me.

How about a compromise: Add BuilderBots, but also add an optional neutron gun for Grabbers, enabled by default until we rework the missions. And make two code battle levels, one with buikding grabber, and one with two starting bots.

**Comment by RaptorParkowsky** [comment]
I also don't remind any of the "main idea" of the Code Battles. They have been never finished anyway since GOLD 0.1.6 and we left them that way as a startup feature, because there's a lot of more possibilities to do with them. And building Grabbers was certainly a placeholder, because we were slowly planning implement Builders for the past ~2 or 3 years and then after that upgrade the Code Battles.

**Comment by MrSimbax** [comment]
```
How about a compromise: Add BuilderBots, but also add an optional neutron gun for Grabbers, enabled by default until we rework the missions. And make two code battle levels, one with buikding grabber, and one with two starting bots.
```

I'm ok with that solution for now. Although in the future we should just add some new object which can build and grab things. Something to think about for our 3D designers.

**Comment by tomangelo2** [comment]
I don't think Grabbers are good as they're. I'm not against BuilderBots as they are. 
I would like to see them in game. They'll be great addition for custom maps. For CodeBattles it looks like we don't know what we thought few years ago when we were implementing this mode, if we'll choose "golden" rules for this mode then there is a problem with limitation by them, in case of ceebot rules they are good change making some new strategical mechanics, so I'm generally positive about them, it's just matter of allowing them by rules. 
I'm not sure however how they would fit into Missions as these bots just weren't there originally, and there were already much discusses about changes in these levels. Still, disabling them would deny possibility to finish game only by programming without building by Grabbers. But then as mentioned before, Grabbers were only placeholder, it looks kinda weird. Then again, we're moving dangerously close to limits of remake.
Last sentence doesn't apply only to Grabbers, I'm kinda worried about another planned features like new build-in languages or level changes (I'm not against them, to be clear, I just wonder how far we can go as a 'remake' + sentence cited before about what we won't implement). I want to avoid something like "You told you won't make so much changes but you made so much changes 0mg w0rst game evah" etc.

**Comment by Emxx52** [comment]
@tomangelo2 @MrSimbax 
Since changing the plot is out of the question I don't plan on adding Builders to Missions.
I thought about including them in our form of New Game+, i.e. the programmer mode we discussed in the past, as a special research or outright replacement for the Astronaut.
It would be really easy to implement this mode in the near future, as we already have level difficulty settings in dev-buzzingcars.

@krzys-h 
Your comment is ridiculous. So all past placeholder solutions are now features?

**Comment by MrSimbax** [comment]
> I thought about including them in our form of New Game+, i.e. the programmer mode we discussed in the past, as a special research or outright replacement for the Astronaut.

++

**Comment by tomangelo2** [comment]
I agree, separate mode is IMHO fair compromise, but I would make it as another mode in main menu, next to normal Mission mode, for difficult levels I would go with different resources/aliens amount, damage multipliers, making buildings unusable when damaged beyond some level, virus lifetime etc. Same would work in Missions+ mode, making it a bit more challenging to code.

**Comment by krzys-h** [comment]
> Since changing the plot is out of the question I don't plan on adding Builders to Missions.
I thought about including them in our form of New Game+, i.e. the programmer mode we discussed in the past, as a special research or outright replacement for the Astronaut.

A programmer mode is what I was thinking about since the beginning, I guess I should have written that more clearly.

> It would be really easy to implement this mode in the near future, as we already have level difficulty settings in dev-buzzingcars.

Adding the code that handles difficulty in dev-buzzingcars is literally 15 minutes of work. If that were to be what stops us from starting to "rework the missions" to handle BuilderBots and other code-playthrough features better, then they would be finished long ago.

> Your comment is ridiculous. So all past placeholder solutions are now features?

Placeholder, by definition, is something that exists for a short period of time while we are working on a replacement. If it exists in the game for 4 years, it's no longer a placeholder.

I wouldn't apply that to ALL past placeholder solutions, but in this case there is already some other things built on top of the assumption that grabbers can build, like code battles and all the programs using build() that were written so far.

**Comment by MrSimbax** [comment]
> Placeholder, by definition, is something that exists for a short period of time while we are working on a replacement. If it exists in the game for 4 years, it's no longer a placeholder.

I don't agree with this definition of placeholder and it seems that most dictionaries also disagree. It doesn't matter how long the placeholder exists, it is still something that can (and probably should) be replaced later, where later doesn't specify if it's 4 years later or 1 hour later.

Just because something exists for a long time, does not mean it's good, and it's definitely the case with Grabber having additional functionality, which is not reflected by its model and role in the original game and earlier version of GOLD.

> but in this case there is already some other things built on top of the assumption that grabbers can build, like code battles and all the programs using build() that were written so far.

We're still in alpha stage and we'll long be seeing in what state the game is right now. We said that things will change a lot. We did not promise to keep anything as it is or as it was. That would be ridiculous and ill for the project.

If I wanted to code in WinAPI (Colobot) I would code in WinAPI, but I want to code in WinForms (GOLD).

**Comment by krzys-h** [comment]
I was hoping we could end this shitposting now that we have agreed on a solution but it looks like you really want to go on :stuck_out_tongue:

> I don't agree with this definition of placeholder and it seems that most dictionaries also disagree. It doesn't matter how long the placeholder exists, it is still something that can (and probably should) be replaced later, where later doesn't specify if it's 4 years later or 1 hour later.

Well, I obviously didn't mean a dictionary definition of a placeholder. It probably comes down to a fact that I actually coded a few bug things using build() already and you didn't. I made [a program that builds a base from nothing with just one robot](https://github.com/krzys-h/colobot-superbase) right after build() was implemented. I made [the 'official' code battle opponent](https://github.com/colobot/colobot-data/blob/master/levels/battles/chapter001/level002/program.txt). And there are other people who [made programs and user levels that would be completly broken by removing this 'feature'](https://colobot.info/forum/forumdisplay.php?fid=60).

Is it good and logical for Grabbers to be able to build? No, it isn't, but there are a lot of illogical things in this game, like robots disappearing during transit between planets for example. Should that mean that we can just remove a feature that was in the game for 4 years? I don't think so, you could equally say that we can remove Sniffers because they are rarely used and it doesn't matter that they were in the game since the original version.

> Grabber having additional functionality, which is not reflected by its model [...]

Who said we cannot update the model?

> We're still in alpha stage and we'll long be by seeing in what state the game is right now. We said that things will change a lot.

I feel like you are just trying to find any argument to convince me now. I thought we all agreed that the game is mostly stable now and we really should stop using the 'alpha' prefix? The only reason we are keeping it really is to have some excuse in case of minor fuckups like release with not working code battles.

I'm not saying the game will not change anymore, but like I've said on many occasions, because of the above I always try to keep backwards compatibility with old features whenever possible. Maybe it doesn't matter considering how small the community of this game is, but it really would if it was bigger.

---

So now, since you apparently REALLY want to keep playing this game: Show me where it says that builders are supposed to be a placeholder BEFORE this functionality was implemented. And I assure you, it's not in #188, #189, #190. The original author of this PR probably never even meant for them to be ever replaced with builders.

**Comment by MrSimbax** [comment]
> I was hoping we could end this shitposting now that we have agreed on a solution

I wAs HOpiNG wE cOUld EnD thIS ShItPOsTing

spongebob mocking

I'm sorry, I could not resist myself, I just love this meme. Anyway, my god, two different point of views and it's already called shitposting. My previous post wasn't even aggressive in any way.

> It probably comes down to a fact that I actually coded a few bug things using build() already and you didn't.

So? I would be more than happy to update my old programs if they were worth updating. I would be aware that the game will probably look a little different in later releases. I don't know who wouldn't. I'm pretty sure that even fully released games can change sometimes something that breaks mods. Should it stop them from changing the game for the better? I think not.

I think you have mindset similar to C++ designers: let's keep the old ways of C for the sake of backwards compatibility. I have more of a Python mindset: let's make major changes so the language is better and stop supporting the old ways at some point.

> Should that mean that we can just remove a feature that was in the game for 4 years?

Yes, yes we can, because it is placeholder and we're in alpha and the progress of this project is so slow I'd say the 4 years you're again talking about would be at most a year in a normally paced project.

And it's not feature removal, quite the opposite, it's an _enhancement_. Every robot in this game has **one** special feature, no less, no more. Builders are made for building, and Grabbers are made for moving stuff. Obviously they were not supposed to build. This is finishing of the yet undone feature of "build()" for Grabbers done years ago for Grabbers.

Let's be honest here, alright? Grabbers can build only because there was no easy way to implement Builders at that point. It caused building Grabbers to appear even in an official teaser, that's a shame and long-term consequence of our slow tempo and keeping placeholders for too long. This is a mistake of the past that we should fix as soon as possible, not defend it and keep it. I feel like you just want to call this bug a feature because you just got used to it and don't want to see more changes that fixing this implies for some reason.

> Who said we cannot update the model?

Judging by Slack, I think that a lot of people will not like this solution any more than a new object as this "BuilderGrabber".

> I feel like you are just trying to find any argument to convince me now.

What's wrong with this argument? I wasn't thinking really hard about it, maybe it's weak, I don't know.

> I thought we all agreed that the game is mostly stable now and we really should stop using the 'alpha' prefix?

Stability is the last thing anyone expects from alpha, and beta for that matter. The most important thing that separates alphas, betas and final releases are features. And, in terms of features, I wouldn't say that we're done here, as many of the new ones were not really thought out fully and properly and need more polishing and design decisions. Call it beta if you want, it doesn't change anything in this argument, just replaces the word "alpha" with "beta".

> because of the above I always try to keep backwards compatibility with old features whenever possible.

Like I said earlier, C++ vs Python mindset. I don't think we'll ever agree on this matter.

> So now, since you apparently REALLY want to keep playing this game:

Ah, I think I know what you're referring to. It's hard to _not_ know what's happening around this game when you are friends with people who are in the middle of it all. Sometimes I just feel like I want to help and/or step in. It's not a crime.

> Show me where it says that **builders** are supposed to be a placeholder BEFORE this functionality was implemented.

I'll assume you meant grabbers.

> And I assure you, it's not in #188, #189, #190. The original author of this PR probably never even meant for them to be ever replaced with builders.

I guess we'll not know what the author thought unless we ask him, but your guess is as good as mine, and I could easily say that the author of this pull request used Grabbers because there was no Builders at that point. Also, don't you think that letting strangers who just made a few pull requests 4 years ago to decide about the final design of the game is bad idea?

I won't be searching for any sources of the word "placeholder" because I don't see it as necessary and I'm wasting my time too much already. I thought that it was obvious for everyone that Grabbers ability to build is just a placeholder for builders. Apparently, I was wrong, since I now realized that you just said in one of the comments that all bots should be able to build, and that's beyond ridiculousness for me, sorry. Wait, why am I saying sorry? Oh, so people don't take my posts as shitposting just because I don't agree with them... Apparently I seem to have the point in discussions since I seem to make the other part mad for no reason too often.

**Comment by krzys-h** [comment]
>> because of the above I always try to keep backwards compatibility with old features whenever possible.
>
> Like I said earlier, C++ vs Python mindset. I don't think we'll ever agree on this matter.

And this probably sums up this whole discussion. And I think that the compromise of allowing both for purposes of userlevels and defaulting to building grabbers *until levels are modified* should satisfy both of these mindsets, so why are we even still discussing this?

> I thought that it was obvious for everyone that Grabbers ability to build is just a placeholder for builders. Apparently, I was wrong, since I now realized that you just said in one of the comments that all bots should be able to build, and that's beyond ridiculousness for me, sorry.

It does make a lot of sense if we are working with the assumption that we don't want to introduce features like new bots into the remake, which we are clearly no longer doing as @tomangelo2 noticed.

> I won't be searching for any sources of the word "placeholder" because I don't see it as necessary and I'm wasting my time too much already.

Well, I did (but I have no idea why am I wasting my time on this either).

The idea for BuilderBots first appeared on IRC on 30th of March 2013 from @Emxx52 as a joke when talking about importing some exercises from CeeBot:
```
1364650250.8 #colobot Simbax: Więc dodanie misji z Ceebota może być problematyczne
1364650250.8 #colobot Emxx52: repeat, build, factory
1364650250.8 #colobot Emxx52: tylko to
1364650250.8 #colobot Simbax: No to mówię, że nowe
1364650250.8 #colobot Emxx52: zawsze mozna przerobic/wywalic te cwiczenia
1364650250.8 #colobot Emxx52: build i factory byly niewykorzystywane
1364650335.84 #colobot Simbax: No w sumie
1364650335.84 #colobot Emxx52: przerobi sie, zeby zamiast repeat bylo for
1364650335.84 #colobot Simbax: Repeat to takie ubogie for
1364650335.84 #colobot Emxx52: i done...
1364650335.84 #colobot Emxx52: a build sie da Builderom :D
1364650406.33 #colobot Emxx52: WheeledBuilder
1364650410.84 #colobot Emxx52: :D
```

And the next time it appeared, apparently we are already going to be implementing that idea because somebody implemented build() in a pull request independently from that discussion?
```
1367089544.46 #colobot krzys_h: Próbujemy sprawdzić czemu build() od XienDev się buguje
[...]
1367089620.2 #colobot Emxx52: Lulz, przecież i tak powiedziałem mu że na kod czekamy, a on nadal swoje. Poza tym nie wprowadzamy możliwości budowania jak w CeeBocie 4, tylko będą Buildery
1367089626.6 #colobot Emxx52: Więc jaki sens?
1367089657.65 #colobot tomangelo: Komenda gotowa do zaimplementowania do botów
1367089664.8 #colobot Raptor: Tak, tylko ciężko będzie zrobić nowy typ bota z wyposażeniem
1367089737.58 #colobot Emxx52: Damy radę
1367089737.58 #colobot tomangelo: Podczas gdy botów jeszcze nie ma, to pracujemy już nad nową komendą do nich
1367089737.58 #colobot tomangelo: Równoległość wykonywania zadań
[...]
1367089784.27 #colobot Emxx52: Wystarczy przerobić builda z CeeBota 4 jak kod dostaniemy, żeby się odpalal tylko na Builderach i ew. ludziach
[...]
1367089868.67 #colobot Raptor: Apropo BuilderBotów. Jak je wpasujemy do gry? W fabule misji COLOBOT nie znajdzie się dla niego miejsca raczej, chyba że dodamy naprawdę dużo misji
1367089868.67 #colobot Raptor: Będą co najwyżej gościć w Free i Ćwiczeniach raczej, nie?
1367089868.67 #colobot tomaszkax86: Raptor: tak samo jak dla Exchange Postów, po prostu nie będzie builder botów w fabule głównej
[...]
1367089933.21 #colobot Emxx52: Raptor: w misjach raczej nie, ale sama możliwość ich budowania gdzieś tam później dobra by była. No i oczywiście Retro
1367089933.22 #colobot tomangelo: Będą w ćwiczeniach i custom levelach
```

At that point we were going to make build() work only for builders. But later we realized (or, to be more precise, I actually tried it) that in the current form it allows you to write a code-only solution to some of the existing missions, and that it's quite a cool feature to make missions that you can finish using code only.

So like I said, I have nothing against builders, but until we introduce a new gamemode for code-only play, I think the 'placeholder' should stay. And it should stay available for userlevels afterwards.

**Comment by Emxx52** [comment]
>The idea for BuilderBots first appeared on IRC on 30th of March 2013 from @Emxx52 as a joke when talking about importing some exercises from CeeBot

'as a joke'? Excuse me? I was just happy to share the idea, I was _clearly_ not joking mate.

>But later we realized (or, to be more precise, I actually tried it) that in the current form it allows you to write a code-only solution to some of the existing missions, and that it's quite a cool feature to make missions that you can finish using code only.

_You_ realized that, and so you refused to implement Builders yourself. Now you try to block my fix for the situation that we have now (i.e. a gamemode that never made any sense).

As it turns out, I should've done all this five years ago. Quite frankly it would've been just as easy to add them back then.

Also: There already is a TODO list in this PR's description, so I have no idea why you repeat the things that are already there.

I'm not going to let Grabbers build again, I won't tolerate a 'placeholder feature', I'm not OK with letting this project become a total mess.
As for the Code Battles, well, I was going to to this either way, because it seems as though you'd rather waste your time on a pointless discussion instead of fixing them yourself.

**Comment by Emxx52** [comment]
Also, this is one of our earliest Facebook posts and it may have been interpreted as a teaser for what's finally done:
https://www.facebook.com/colobotgame/photos/a.551708611518873.1073741825.434620166561052/558977940791940/?type=3&theater

**Comment by krzys-h** [comment]
Okay, maybe 'joke' was not the best word to use. I meant a randomly shared idea that I can't trace how it became an official planned feature. And like I said, that doesn't mean I'm against it.

I'm not trying to block Builders, if that was my intention I'd have tried to stop it long ago. I'm just trying to prevent something that accidentaly became an interesting idea from being erased until we finally implement programming-only missions (which is also a feature we discussed and wanted to implement), even if it barely forks in the current state.

I put the TODO list again because 1) GitHub requires a comment when changing state of pull request to 'changes requested' 2) To sum up what I thought we already agreed on...

And you can't say "I won't tolerate building grabbers" just like I can't say "I won't tolerate non-object-oriented factory()". You have to convince me that it's really better to get rid of a working feature, and I really don't think it is.

**Comment by krzys-h** [comment]
@Emxx52 so if I make a meme with Pinkie Pie on the moon then it becomes a teaser of a feature we're going to implement?

**Comment by Emxx52** [comment]
The thing is, it was _never_ considered a feature up until this discussion. In fact, noone but yourself _does consider_ it a feature. That's why I said I won't tolerate keeping this placeholder intact. If I were to do that, that would also mean I'm OK with CBOT beeing all messed up, so what would be the point of dev-cbot-factory?

**Comment by krzys-h** [comment]
To be clear: *feature* in the last post was not meant to mean *building grabbers*, it was meant to mean *passing missions with code*. And I can't agree with the statement that it's not a featire.

**Comment by tomangelo2** [comment]
> because it seems as though you'd rather waste your time on a pointless discussion instead of fixing them yourself.

Lets everyone fix everything with his idea and later think about why every part of the game is different than the rest of it. @krzys-h will fix everything with his idea, @Emxx52 with his idea, @RaptorParkowsky will make commits with his idea and everything will be one big mess. District breakdown isn't good at the end.

**Comment by krzys-h** [comment]
Okay, I tested this and:
* The Builder research takes way too long. I'm not sure how research length calculation works, but the harder researches take longer to process, and if you want this to be one of first researches definitely shouldn't take that long
* The new ReseachCenter UI seems slightly off-center - in reality it's probably not, but it looks like that because of the spacing from the self-destruct button
* The Builder UI works way better than I thought it would :+1:

**Comment by rbcat** [comment]
Sorry for barging it, it appears that I am way, way too late for party. But how about making this feature switchable within scene.txt, for back compatibility? I mean, adding global parameter defining whether 0) builder will exist and grabbers will not be able to build or 1) there will be no builders and grabbers will still able to build. By default (if parameter is NOT specified) it will be 1, so it will NOT automatically fracture unchanged levels including those created by users times before. You on your side will just have to set it to 0 within original mission levels as well as those who will be creating their own levels and wishing to use builders. It doesn't looks like very aesthetic solution and perhaps it might contradict whole philosophy of Colobot Gold development, but at least it should be safer.

File: src/object/task/taskbuild.cpp [hunk 1]
```diff
@@ -392,15 +411,17 @@ Error CTaskBuild::Start(ObjectType type)
     err = FlatFloor();
     if ( err != ERR_OK )  return err;
 
+    pv = m_object->GetPosition();
+    pm = m_metal->GetPosition();
+    if(!m_physics->GetLand() && abs(pm.y-pv.y)>8.0f) return ERR_BUILD_METALAWAY;
+
```
Comment (MrSimbax) [inline src/object/task/taskbuild.cpp:418]: Linter issue: `using integer absolute value function 'abs' when argument is of floating point type`

**Pull request merged**